src/routes/webhooks/customerio.ts

```diff
@@ -0,0 +1,98 @@
+import { createHmac, timingSafeEqual } from 'crypto';
+import { FastifyInstance, FastifyRequest } from 'fastify';
+import createOrGetConnection from '../../db';
+import { UserMarketingCta } from '../../entity';
+import { cachePrefillMarketingCta } from '../../schema/users';
+import { logger } from '../../logger';
+
+const verifyCIOSignature = (
+  webhookSigningSecret: string,
+  req: FastifyRequest,
+): boolean => {
+  const timestamp = req.headers['x-cio-timestamp'] as string;
+  const signature = req.headers['x-cio-signature'] as string;
+
+  if (!timestamp || !signature) {
+    return false;
+  }
+
+  const hmac = createHmac('sha256', webhookSigningSecret);
+  hmac.update(`v0:${timestamp}:`);
+  hmac.update(req.rawBody);
+
+  const hash = hmac.digest();
+  if (!timingSafeEqual(hash, Buffer.from(signature, 'hex'))) {
+    logger.debug("CIO Signature didn't match");
+    return false;
+  }
+
+  logger.debug('CIO Signature matched!');
+  return true;
+};
+
+type MarketingCtaPayload = {
+  Body: {
+    userId: string;
+    marketingCtaId: string;
+  };
+};
+
+export const customerio = async (fastify: FastifyInstance): Promise<void> => {
+  fastify.addHook<MarketingCtaPayload>('preValidation', async (req, res) => {
+    const valid = verifyCIOSignature(process.env.CIO_WEBHOOK_SECRET, req);
+    if (!valid) {
+      return res.status(403).send({ error: 'Invalid signature' });
+    }
+  });
+
+  fastify.post<MarketingCtaPayload>('/marketing_cta', {
+    config: {
+      rawBody: true,
+    },
+    handler: async (req, res) => {
+      try {
+        const { userId, marketingCtaId } = req.body;
+
+        const con = await createOrGetConnection();
+        await con.getRepository(UserMarketingCta).upsert(
+          {
+            userId: userId,
+            marketingCtaId: marketingCtaId,
+          },
+          ['userId', 'marketingCtaId'],
+        );
+
+        await cachePrefillMarketingCta(con, userId);
```

Comment:
This will give a edge-case right?
EG user gets added to two campaigns without viewing one he'll get auto assigned the later in redis, not the first?

Reply:
`cachePrefillMarketingCta` is the same internal function as the boot request uses. I.e. it selects from database the first in queue (there's a test for it)

Reply:
I.e. you've been assigned to campaign A, first which loads A into cache, then get assigned to campaign B, which then loads A into cache because A was first in the queue (based in createdAt)

Reply:
https://github.com/dailydotdev/daily-api/blob/b067f758fc865b0c92cbc4c4fdea437bd9d9f869/src/schema/users.ts#L764-L774